src/coreclr/System.Private.CoreLib/src/System/Array.CoreCLR.cs

```diff
@@ -252,8 +268,47 @@ private unsafe nint GetFlattenedIndex(ReadOnlySpan<int> indices)
             }
         }
 
-        [MethodImpl(MethodImplOptions.InternalCall)]
-        internal extern unsafe object? InternalGetValue(nint flattenedIndex);
+        internal unsafe object? InternalGetValue(nint flattenedIndex)
+        {
+            MethodTable* pMethodTable = RuntimeHelpers.GetMethodTable(this);
+
+            TypeHandle arrayElementTypeHandle = pMethodTable->GetArrayElementTypeHandle();
+
+            // Legacy behavior (none of the cases where the element type is a type descriptor are supported).
+            // This includes the case where the element type is a pointer or a function pointer too.
+            if (arrayElementTypeHandle.IsTypeDesc)
+            {
+                ThrowHelper.ThrowNotSupportedException(ExceptionResource.Arg_TypeNotSupported);
+            }
+
+            Debug.Assert((nuint)flattenedIndex < NativeLength);
+
+            ref byte arrayDataRef = ref Unsafe.AddByteOffset(ref Unsafe.As<RawData>(this).Data, pMethodTable->BaseSize - (nuint)(2 * sizeof(IntPtr)));
```

Comment:
```suggestion
            ref byte arrayDataRef = ref MemoryMarshal.GetArrayDataReference(this);
```

Reply:
Just to understand this better, I had inlined that logic there since we already had the `MethodTable*` pointer here, so it seemed like it could potentially skip and entra indirection to get it again from `MemoryMarshal.GetArrayDataReference`. Is that not the case and the JIT can potentially see that and optimize that away anyway, or is that just not worth the extra complexity here? I'll change this but figure I'd ask 🙂

Reply:
The JIT should be smart enough to CSE the extra MethodTable fetch if it is worth it.

We do not want this magic formula to be copy&pasted in too many places.

Reply:
I see, yeah that makes perfect sense. Thank you! 😊